transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java

```diff
@@ -381,28 +353,35 @@ protected void run() {
                             wakenUp.set(0);
                         }
                         if (!hasTasks()) {
-                            strategy = epollWait();
+                            // When we are in the EventLoop we don't bother setting the timerFd for each
+                            // scheduled task, but instead defer the processing until the end of the EventLoop
+                            // (next wait) to reduce the timerFd modifications.
+                            timerFdDeadline = checkScheduleTaskQueueForNewDelay(timerFdDeadline);
+                            try {
+                                strategy = epollWait();
+                            } finally {
+                                // This getAndAdd will change the raw value of nextDeadlineNanos to be negative
+                                // which will block any *new* timerFd mods by other threads while also "preserving"
+                                // its last value to avoid disrupting a possibly-concurrent setTimerFd call
+                                // (so that we can know the timerFd really did/will get updated to the read value).
+                                timerFdDeadline = nextDeadlineNanos.getAndAdd(Long.MAX_VALUE + 1);
+                                // The value of nextDeadlineNanos is now guaranteed to be negative
+                            }
                         }
                         // fallthrough
                     default:
                 }
 
                 try {
-                    processReady(events, strategy);
-                } finally {
-                    try {
-                        // Note the timerFd code depends upon running all the tasks on each event loop run. This is so
-                        // we can get an accurate "next wakeup time" after the event loop run completes.
-                        runAllTasks();
-                    } finally {
-                        // No need to drainScheduledQueue() after the fact, because all in event loop scheduling results
-                        // in direct addition to the scheduled priority queue.
-
-                        // When we are in the EventLoop we don't bother setting the timerFd for each scheduled task, but
-                        // instead defer the processing until the end of the EventLoop to reduce the timerFd
-                        // modifications.
-                        checkScheduleTaskQueueForNewDelay();
+                    boolean timerFdFired = processReady(events, strategy);
+                    if (timerFdFired) {
```

Comment:
just merge the above lines ?

Reply:
Sure, I had done it like this to be more "self documenting", but can add a comment instead